Throwing the GCC static analyzer at it with `-fanalyzer` yields the following:

```
gcc -DHAVE_CONFIG_H -I.    -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -Wall -Wextra -fanalyzer -MT jma.o -MD -MP -MF .deps/jma.Tpo -c -o jma.o jma.c
jma.c: In function ‘read_header’:
jma.c:42:9: warning: ‘strncpy’ output truncated before terminating nul copying 2 bytes from a string of the same length [-Wstringop-truncation]
   42 |         strncpy(header.basin, "WP", 2);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Any of the options at https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wstringop-truncation is acceptable as a fix.

--- a/tracks/atcf.c
+++ b/tracks/atcf.c
@@ -104,9 +104,9 @@ struct stormdata *read_stormdata_atcf(struct stormdata *storms,
 					if (strlen(token[27]) < STORM_NAME_LENGTH) {
 						storm.header.name[strlen(token[27])] = '\0';
 					}
-					storm.header.id = atoi(token[1]);
 				}
 			}
+			storm.header.id = atoi(token[1]);
 		} else {
 			strncpy(storm.header.name, "UNNAMED", 8);
 			storm.header.id = atoi(token[1]);

--- a/tracks/atcf.c
+++ b/tracks/atcf.c
@@ -30,7 +30,7 @@ static char *stripspace(char *str)
 }
 
 struct stormdata *read_stormdata_atcf(struct stormdata *storms,
-									  struct storm_arg *args)
+									  struct storm_arg *args, bool skipasynoptic)
 {
 	FILE *file;
 	int i;
@@ -39,7 +39,7 @@ struct stormdata *read_stormdata_atcf(struct stormdata *storms,
 	char *token[35];
 	int dateset = 0;
     int points = 0;
-	int lasttype;
+	int lasttype = 0;
 	struct pos pos;
 	file = fopen(args->input, "r");
 	
@@ -65,45 +65,52 @@ struct stormdata *read_stormdata_atcf(struct stormdata *storms,
 			i++;
 		}
         
-        if (i < 12) {
+        if (i < 9) {
             fprintf(stderr, "Parsing error in file '%s': Not enough data fields to parse data point properly\n",
                     args->input);
             exit(-1);
         }
-
-		/* Begin storing data */
-		/* Check the storm's characteristics for this time point */
-		if ((strcasecmp(token[10], "TD") == 0) || (strcasecmp(token[10], "TS") == 0) || (strcasecmp(token[10], "TY") == 0) || 
-			(strcasecmp(token[10], "ST") == 0) || (strcasecmp(token[10], "TC") == 0) || (strcasecmp(token[10], "HU") == 0)) {
-			pos.type = TROPICAL;
-			lasttype = TROPICAL;
-		} else if ((strcasecmp(token[10], "SD") == 0) || (strcasecmp(token[10], "SS") == 0)) {
-			pos.type = SUBTROPICAL;
-			lasttype = SUBTROPICAL;
-		} else if ((strcasecmp(token[10], "EX") == 0)) {
-			pos.type = EXTRATROPICAL;
-			lasttype = EXTRATROPICAL;
-		} else if ((strcasecmp(token[10], "LO") == 0) || (strcasecmp(token[10], "WV") == 0) || (strcasecmp(token[10], "MD") == 0) || (strcasecmp(token[10], "DB") == 0)) {
-			pos.type = LOW;
-			lasttype = LOW;
-		} else {
-			/* If we have a different condition (e.g. inland, dissipating), 
-			 * assume that the last valid condition is still true */
-			pos.type = lasttype;
-		}
-		
-		/* Set the storm name and ID */
-		if ((strcasecmp(token[10], "TS") == 0) || (strcasecmp(token[10], "TY") == 0) || (strcasecmp(token[10], "ST") == 0) || 
-			(strcasecmp(token[10], "TC") == 0) || (strcasecmp(token[10], "HU") == 0) || (strcasecmp(token[10], "SS") == 0)) {
-			if (strcasecmp(token[27], storm.header.name) != 0) {
-				strncpy(storm.header.name, token[27], strlen(token[27]));
-				if (strlen(token[27]) < STORM_NAME_LENGTH) {
-					storm.header.name[strlen(token[27])] = '\0';
-				}
-				storm.header.id = atoi(token[1]);
+		if (i >= 11) {
+			/* Begin storing data */
+			/* Check the storm's characteristics for this time point */
+			if ((strcasecmp(token[10], "TD") == 0) || (strcasecmp(token[10], "TS") == 0) || (strcasecmp(token[10], "TY") == 0) || 
+				(strcasecmp(token[10], "ST") == 0) || (strcasecmp(token[10], "TC") == 0) || (strcasecmp(token[10], "HU") == 0)) {
+				pos.type = TROPICAL;
+				lasttype = TROPICAL;
+			} else if ((strcasecmp(token[10], "SD") == 0) || (strcasecmp(token[10], "SS") == 0)) {
+				pos.type = SUBTROPICAL;
+				lasttype = SUBTROPICAL;
+			} else if ((strcasecmp(token[10], "EX") == 0)) {
+				pos.type = EXTRATROPICAL;
+				lasttype = EXTRATROPICAL;
+			} else if ((strcasecmp(token[10], "LO") == 0) || (strcasecmp(token[10], "WV") == 0) || (strcasecmp(token[10], "MD") == 0) || (strcasecmp(token[10], "DB") == 0)) {
+				pos.type = LOW;
+				lasttype = LOW;
+			} else {
+				/* If we have a different condition (e.g. inland, dissipating), 
+			 	* assume that the last valid condition is still true */
+				pos.type = lasttype;
 			}
+		} else {
+			fprintf(stdout, "Not enough fields to infer storm type; falling back to tropical.\n");
+			pos.type = TROPICAL; //Fallback in case of shortened ATCF
 		}
-		
+		/* Set the storm name and ID */
+		if (i >= 11) {
+			if ((strcasecmp(token[10], "TS") == 0) || (strcasecmp(token[10], "TY") == 0) || (strcasecmp(token[10], "ST") == 0) || 
+				(strcasecmp(token[10], "TC") == 0) || (strcasecmp(token[10], "HU") == 0) || (strcasecmp(token[10], "SS") == 0)) {
+				if (strcasecmp(token[27], storm.header.name) != 0) {
+					strncpy(storm.header.name, token[27], strlen(token[27]));
+					if (strlen(token[27]) < STORM_NAME_LENGTH) {
+						storm.header.name[strlen(token[27])] = '\0';
+					}
+					storm.header.id = atoi(token[1]);
+				}
+			}
+		} else {
+			strncpy(storm.header.name, "UNNAMED", 8);
+			storm.header.id = atoi(token[1]);
+		}
 		if (!dateset) {
             strncpy(storm.header.basin, token[0], 2);
             storm.header.basin[2] = '\0';
@@ -119,7 +126,9 @@ struct stormdata *read_stormdata_atcf(struct stormdata *storms,
         pos.month = substr_to_int(token[2], 4, 2);
         pos.day = substr_to_int(token[2], 6, 2);
         pos.hour = substr_to_int(token[2], 8, 2);
-		
+	if (skipasynoptic && (pos.hour % 6 != 0)) {
+		continue;
+	}
 		/* latitude */
 		if (token[6][strlen(token[6])-1] == 'N') {
 			pos.lat = (double)atoi(token[6]) / 10.0;

--- a/tracks/atcf.h
+++ b/tracks/atcf.h
@@ -11,6 +11,6 @@
 #include "track.h"
 
 struct stormdata *read_stormdata_atcf(struct stormdata *storms,
-									 struct storm_arg *args);
+									 struct storm_arg *args, bool skipasynoptic);
 
 #endif

--- a/tracks/hurdat2.c
+++ b/tracks/hurdat2.c
@@ -30,7 +30,8 @@ static char *stripspace(char *str)
 }
 
 struct stormdata *read_stormdata_hurdat2(struct stormdata *storms,
-                                         struct storm_arg *args)
+                                         struct storm_arg *args,
+				         bool skipasynoptic)
 {
 	FILE *file;
 	int i;
@@ -125,7 +126,6 @@ struct stormdata *read_stormdata_hurdat2(struct stormdata *storms,
             /* hour */
             pos.hour = substr_to_int(token[1], 0, 2);
             minute = substr_to_int(token[1], 2, 2);
-            
             /* Check whether this is a synoptic time */
             if ((pos.hour % 6 == 0) && minute == 0) {
                 /* This is a synoptic time */
@@ -204,7 +204,7 @@ struct stormdata *read_stormdata_hurdat2(struct stormdata *storms,
             /* NOTE: We are skipping non-synoptic times right now, 
              * until we figure out what to do with them! 
              */
-            if (syn_time)
+            if (syn_time || !skipasynoptic)
                 save_pos(args, storms, &storm, &pos);
 
             

--- a/tracks/hurdat2.h
+++ b/tracks/hurdat2.h
@@ -11,6 +11,6 @@
 #include "track.h"
 
 struct stormdata *read_stormdata_hurdat2(struct stormdata *storms,
-									 struct storm_arg *args);
+									 struct storm_arg *args, bool skipasynoptic);
 
 #endif

--- a/tracks/jma.c
+++ b/tracks/jma.c
@@ -39,7 +39,7 @@ static struct storm_header read_header(char** token)
 	header.id = identifier % 100;
 	header.year = get_full_year(identifier/100);
 	strncpy(header.name, token[7], strlen(token[7]));
-	strncpy(header.basin, "WP", 2);
+	strncpy(header.basin, "WP", 3);
 	return header;
 }
 struct stormdata *read_stormdata_jma(struct stormdata *storms, struct storm_arg* args)

--- a/tracks/jma.c
+++ b/tracks/jma.c
@@ -42,7 +42,7 @@ static struct storm_header read_header(char** token)
 	strncpy(header.basin, "WP", 3);
 	return header;
 }
-struct stormdata *read_stormdata_jma(struct stormdata *storms, struct storm_arg* args)
+struct stormdata *read_stormdata_jma(struct stormdata *storms, struct storm_arg* args, bool skipasynoptic)
 {
 	FILE* file;
 	int i;
@@ -99,6 +99,9 @@ struct stormdata *read_stormdata_jma(struct stormdata *storms, struct storm_arg*
 			pos.month = (date % 1000000)/10000;
 			pos.day = (date % 10000)/100;
 			pos.hour = (date % 100);
+			if (skipasynoptic && (pos.hour % 6 != 0)) {
+				continue;
+			}
 			if (!dateset) {
 				storm.header.year = pos.year;
 				storm.header.month = pos.month;

--- a/tracks/jma.h
+++ b/tracks/jma.h
@@ -8,6 +8,6 @@
 #include "track.h"
 
 struct stormdata *read_stormdata_jma(struct stormdata *storms,
-                                                                         struct storm_arg *args);
+                                                                         struct storm_arg *args, bool skipasynoptic);
 
 #endif

--- a/tracks/track.c
+++ b/tracks/track.c
@@ -78,6 +78,7 @@ static void help(void)
 	 "in the argument list this halts the current storm spec and\n"
 	 "instead adds another storm to the list.\n");
   printf("  --useoldcolorkey		Use the legacy color key when useoldcolorkey≠0\n");
+  printf("  --skipasynoptic             Set to 0 to not have the maker skip asynoptic points, i.e. those whose hours are not a multiple of 6 hours (0z, 6z, 12z, 18z).");
 }
 
 static void init_storm_arg(struct storm_arg *stormp)
@@ -192,7 +193,8 @@ static struct args read_args(const int argc, char **argv)
     .bg = "../data/bg8192.png",
     .output = "../png/output.png",
     .scale = SSHWS_CODE,
-    .useoldcolorkey = 0
+    .useoldcolorkey = 0,
+    .skipasynoptic = 1
   };
 
   args.storm = malloc(sizeof(*args.storm));
@@ -300,6 +302,9 @@ static struct args read_args(const int argc, char **argv)
       } else if (strcasecmp(argv[i], "--useoldcolorkey") == 0) {
 	i++;
 	args.useoldcolorkey = (atoi(argv[i]) != 0);
+      } else if (strcasecmp(argv[i], "--skipasynoptic") == 0) {
+	i++;
+	args.skipasynoptic = (atoi(argv[i]) != 0);
       } else if (strcasecmp(argv[i], "--bg") == 0) {
 	i++;
 	args.bg = argv[i];
@@ -1170,11 +1175,11 @@ int main(int argc, char **argv)
     if (!format || strcasecmp(format, "hurdat") == 0) {
       storms = read_stormdata_hurdat(storms, &args.storm[i]);
     } else if (strcasecmp(format, "atcf") == 0) {
-		storms = read_stormdata_atcf(storms, &args.storm[i]);
+		storms = read_stormdata_atcf(storms, &args.storm[i], args.skipasynoptic);
     } else if (strcasecmp(format, "jma") == 0) {
-		storms = read_stormdata_jma(storms, &args.storm[i]);
+		storms = read_stormdata_jma(storms, &args.storm[i], args.skipasynoptic);
     } else if (strcasecmp(format, "hurdat2") == 0) {
-		storms = read_stormdata_hurdat2(storms, &args.storm[i]);
+		storms = read_stormdata_hurdat2(storms, &args.storm[i], args.skipasynoptic);
     } else if (strcasecmp(format, "md") == 0) {
       storms = read_stormdata_md(storms, &args.storm[i]);
     } else if (strcasecmp(format, "tcr") == 0) {

--- a/tracks/track.c
+++ b/tracks/track.c
@@ -64,7 +64,7 @@ static void help(void)
   printf("  --next			Add on another storm\n");
   printf("  --res 			Set the horizontal resolution of output image\n");
   printf("  --bg Set 			map to use for background\n");
-  printf("  --output 			Set the output file\n");
+  printf("  --output 			Set the output file; is ignored when inferoutputlocation≠0\n");
   printf("  --template 			Set to non-zero to pre-fill {{WPTC track map}} (default: on)\n");
   printf("  --alpha  			Set transparency for storm tracks\n");
   printf("  --extra  			Do not cut off the extratropical portion of the tracks when extra≠0\n");
@@ -79,6 +79,7 @@ static void help(void)
 	 "instead adds another storm to the list.\n");
   printf("  --useoldcolorkey		Use the legacy color key when useoldcolorkey≠0\n");
   printf("  --skipasynoptic             Set to 0 to not have the maker skip asynoptic points, i.e. those whose hours are not a multiple of 6 hours (0z, 6z, 12z, 18z).");
+  printf("  --inferoutputlocation       Automatically infer the file name when inferoutputlocation≠0. For individual storms, <storm name, or designation if no storm name found in file> <year> path.png, or when multiple storms are present, <year> <season type name, such as Atlantic hurricane season> summary.png, where the season type is inferred from the basin of the first storm given.\n");
 }
 
 static void init_storm_arg(struct storm_arg *stormp)
@@ -194,7 +195,8 @@ static struct args read_args(const int argc, char **argv)
     .output = "../png/output.png",
     .scale = SSHWS_CODE,
     .useoldcolorkey = 0,
-    .skipasynoptic = 1
+    .skipasynoptic = 1,
+    .inferoutputlocation = 0
   };
 
   args.storm = malloc(sizeof(*args.storm));
@@ -305,6 +307,9 @@ static struct args read_args(const int argc, char **argv)
       } else if (strcasecmp(argv[i], "--skipasynoptic") == 0) {
 	i++;
 	args.skipasynoptic = (atoi(argv[i]) != 0);
+      } else if (strcasecmp(argv[i], "--inferoutputlocation") == 0) {
+	i++;
+	args.inferoutputlocation = (atoi(argv[i]) != 0);
       } else if (strcasecmp(argv[i], "--bg") == 0) {
 	i++;
 	args.bg = argv[i];
@@ -404,8 +409,10 @@ void save_storm(struct storm_arg *args, struct stormdata *storms, struct storm *
     storms->nstorms++;
     storms->storms = realloc(storms->storms,
 			     storms->nstorms * sizeof(*storms->storms));
+    for (int i = 1; storm->header.name[i]; i++) {
+        storm->header.name[i] = tolower(storm->header.name[i]);
+    }
     storms->storms[storms->nstorms - 1] = *storm;
-
     storms->maxlon = MAX(storm->maxlon, storms->maxlon);
     storms->minlon = MIN(storm->minlon, storms->minlon);
     storms->maxlat = MAX(storm->maxlat, storms->maxlat);
@@ -1041,6 +1048,97 @@ static void print_extra_data(struct stormdata *storms)
   print_strongest_storms(storms);
 }
 
+static void infer_output_location(char* output_location, struct stormdata *storms) {
+        int storm_year = storms->storms[0].header.year;
+	if (storms->nstorms == 1) { // Single storm track
+                if (strlen(storms->storms[0].header.name) > 0) {
+			char* storm_name = storms->storms[0].header.name;
+			snprintf(output_location, strlen(storm_name) + 16, "%s %d path.png", storm_name, storm_year);
+                } else {
+                        if (strlen(storms->storms[0].header.basin) == 0) {
+				fprintf(stderr, "--inferoutputlocation was enabled but no name was found for this storm, and the basin also cannot be inferred.\n");
+				exit(-1);
+			}
+			char storm_name[4];
+			if (strcmp(storms->storms[0].header.basin, "SH") == 0) {
+				if (storms->storms[0].npos == 0) {
+					fprintf(stderr, "--inferoutputlocation was enabled but no name was found for this storm, and because it has no positions and is in the southern hempisphere, it is not possible to infer its basin letter.\n");
+					exit(-1);
+				}
+				if (storms->storms[0].pos[0].lon >= 160.0) {
+					snprintf(storm_name, 3, "%d%c", storms->storms[0].header.id, 'P');
+                                } else {
+					snprintf(storm_name, 3, "%d%c", storms->storms[0].header.id, 'S');
+				}
+			} else if (strcmp(storms->storms[0].header.basin, "AL") == 0) {
+				snprintf(storm_name, 4, "%d%c", storms->storms[0].header.id, 'L');
+			} else if (strcmp(storms->storms[0].header.basin, "EP") == 0) {
+				snprintf(storm_name, 4, "%d%c", storms->storms[0].header.id, 'E');
+			} else if (strcmp(storms->storms[0].header.basin, "CP") == 0) {
+				snprintf(storm_name, 4, "%d%c", storms->storms[0].header.id, 'C');
+			} else if (strcmp(storms->storms[0].header.basin, "IO") == 0) {
+				if (storms->storms[0].npos == 0) {
+					fprintf(stderr, "--inferoutputlocation was enabled but no name was found for this storm, and because there were no positions and the basin is the North Indian Ocean, it cannot be inferred whether A or B should be the basin letter.\n");
+					exit(-1);
+				}
+				if (storms->storms[0].pos[0].lon >= 77.55) {
+					snprintf(storm_name, 3, "%d%c", storms->storms[0].header.id, 'B');
+				} else {
+					snprintf(storm_name, 3, "%d%c", storms->storms[0].header.id, 'A');
+				}
+			} else {
+				fprintf(stderr, "Basin %s is not recognized.\n", storms->storms[0].header.basin);
+				exit(-1);
+			}
+                        snprintf(output_location, strlen(storm_name) + 16, "%s %d path.png", storm_name, storm_year);
+		}
+	} else if (storms->nstorms > 1) { //Avoid == 0
+		char season_name[40];
+		memset(season_name, 0, 40);
+		int basin_name_length = 0;
+		bool is_shem = 0;
+		struct storm storm = storms->storms[0];
+		if (strlen(storm.header.basin) == 0) {
+			fprintf(stderr, "--inferoutputlocation was enabled and there are multiple storms, but the basin cannot be inferred. Please instead pass the output location manually.\n");
+			exit(-1);
+		}
+		if (strcmp(storm.header.basin, "WP") == 0) {
+			strncpy(season_name, "Pacific typhoon season", 23);
+		} else if (strcmp(storm.header.basin, "EP") == 0 || strcmp(storm.header.name, "CP") == 0) {
+			strncpy(season_name, "Pacific hurricane season", 25);
+		} else if (strcmp(storm.header.basin, "AL") == 0) {
+			strncpy(season_name, "Atlantic hurricane season", 26);
+		} else if (strcmp(storm.header.basin, "IO") == 0) {
+			strncpy(season_name, "North Indian Ocean cyclone season", 34);
+		} else if (strcmp(storm.header.basin, "SH") == 0) {
+			if (storm.npos == 0) {
+				fprintf(stderr, "--inferoutputlocation was enabled and there are multiple storms, but the basin is a southern hemisphere one and the first storm has no positions to infer which southern hemisphere basin.\n");
+				exit(-1);
+			}
+			double startinglon = storm.pos[0].lon;
+			is_shem = 1;
+			if (startinglon <= -70.0) {
+				strncpy(season_name, "South-East Pacific cyclone season", 34);
+			} else if (startinglon <= 25.0) {
+			        strncpy(season_name, "South Atlantic hurricane season", 31);
+			} else if (startinglon <= 90.0) {
+				strncpy(season_name, "South-West Indian Ocean cyclone season", 39);
+			} else if (startinglon <= 160.0) {
+				strncpy(season_name, "Australian region cyclone season", 33);
+			} else {
+				strncpy(season_name, "South Pacific cyclone season", 29);
+			}
+		} else {
+			fprintf(stderr, "--inferoutputlocation was enabled and there are multiple storms, but the basin designation %s cannot be recognized.\n", storm.header.basin);
+		}
+		if (is_shem) {
+			int previous_year = storm.header.year - 1;
+			snprintf(output_location, 24 + strlen(season_name), "%d-%d %s summary.png", storm.header.year, previous_year, season_name);
+		} else {
+			snprintf(output_location, 19 + strlen(season_name), "%d %s summary.png", storm.header.year, season_name);		}
+	}
+}
+
 static void write_stormdata(struct stormdata *storms, struct args *args)
 {
   cairo_surface_t *surface;
@@ -1206,10 +1304,15 @@ int main(int argc, char **argv)
     }
     count = storms->nstorms;
   }
-
+  if (args.inferoutputlocation) {
+	args.output = malloc(128);
+	infer_output_location(args.output, storms);
+  }
   write_stormdata(storms, &args);
   print_extra_data(storms);
-
+  if (args.inferoutputlocation) {
+	free(args.output);
+  }
   free_stormdata(storms);
 
   return 0;

--- a/tracks/track.h
+++ b/tracks/track.h
@@ -24,6 +24,7 @@ struct args {
   int scale;
   struct colormap *colors;
   bool useoldcolorkey;
+  bool skipasynoptic;
 };
 
 struct storm_arg {

--- a/tracks/track.h
+++ b/tracks/track.h
@@ -20,11 +20,12 @@ struct args {
   double dots, lines;
   double alpha;
   const char *bg;
-  const char *output;
+  char *output;
   int scale;
   struct colormap *colors;
   bool useoldcolorkey;
   bool skipasynoptic;
+  bool inferoutputlocation;
 };
 
 struct storm_arg {

